src/shared/clients/defaultS3Client.ts

```diff
@@ -0,0 +1,411 @@
+                    MaxKeys: DEFAULT_MAX_KEYS,
+                    Prefix: request.folderPath,
+                    ContinuationToken: request.continuationToken,
+                })
+                .promise()
+        } catch (e) {
+            throw new S3Error(`Failed to list objects for bucket ${request.bucketName}: ${e}`)
+        }
+
+        const files: File[] = _(response.Contents)
+            .reject(file => file.Key === request.folderPath)
+            .map(
+                file =>
+                    new DefaultFile({
+                        key: file.Key!,
+                        partitionId: this.partitionId,
+                        bucketName: request.bucketName,
+                        lastModified: file.LastModified,
+                        sizeBytes: file.Size,
+                    })
+            )
+            .value()
+
+        const folders: Folder[] = _(response.CommonPrefixes)
+            .map(prefix => prefix.Prefix)
+            .compact()
+            .map(path => new DefaultFolder({ path, partitionId: this.partitionId, bucketName: request.bucketName }))
+            .value()
+
+        return {
+            files,
+            folders,
+            continuationToken: response.NextContinuationToken,
+        }
+    }
+
+    /**
+     * Looks up the region for the given bucket
+     *
+     * Note that although there is an S3#GetBucketLocation API,
+     * this is the suggested method of obtaining the region.
+     *
+     * @throws S3Error if there is an error calling S3.
+     */
+    private async lookupRegion(bucketName: string, s3: S3): Promise<string> {
+        try {
+            const response = await s3.headBucket({ Bucket: bucketName }).promise()
+            return response.$response.httpResponse.headers[BUCKET_REGION_HEADER]
+        } catch (e) {
+            throw new S3Error(`Failed to find region for bucket ${bucketName}: ${e}`)
+        }
+    }
+}
+
+export interface FileStreams {
+    createReadStream(uri: vscode.Uri): stream.Readable
+    createWriteStream(uri: vscode.Uri): stream.Writable
+}
+
+export class DefaultBucket implements Bucket {
+    public readonly name: string
+    public readonly region: string
+    public readonly arn: string
+
+    public constructor({ partitionId, region, name }: { partitionId: string; region: string; name: string }) {
+        this.name = name
+        this.region = region
+        this.arn = `arn:${partitionId}:s3:::${name}`
+    }
+
+    public toString(): string {
+        return `Bucket (name=${this.name}, region=${this.region}, arn=${this.arn})`
+    }
+}
+
+export class DefaultFolder implements Folder {
+    public readonly name: string
+    public readonly path: string
+    public readonly arn: string
+
+    public constructor({ partitionId, bucketName, path }: { partitionId: string; bucketName: string; path: string }) {
+        this.path = path
+        this.arn = `arn:${partitionId}:s3:::${bucketName}/${path}`
```

Comment:
Can you de-duplicate the arn generation between DefaultFolder and DefaultFile

Reply:
Done